### DISCUSSION TITLE
How to use getStaticPaths dynamically in next js?

### DISCUSSION THREAD

**Discussion by 1mehdifaraji** [vercel/next.js#37803] [Help]
I have 3 blog categories which I need to paginate static pages for each of them . The problem is the number of pages may vary and I can't get access to anything on what category needs to be fetched in getStaticPaths .

The project folder structure is like this :

<img width="296" alt="Screen Shot 1401-03-28 at 10 23 20" src="https://user-images.githubusercontent.com/63982703/174425149-a6c48d6d-2df7-4862-9ea9-b8bd68b808ee.png">

My code is like this :

```
export const getStaticPaths: GetStaticPaths = async (props) => {
    // const { category } = props.params;  // Cant Access to dynamic category name from the url

    const blogs = await client.getEntries({
      content_type: "blog",
     // "fields.category": category, // Cant Access to dynamic category name from the url
    });

    let pages;
    const howmany = blogs.total / 12;
    pages = Math.ceil(howmany / 1) * 1;
    let paths = [];
    for (var i = 0; i < pages; i++) {
      paths.push({
        params: { page: `${i + 1}` },
      });
    }

  return {
    paths ,
    fallback: "blocking",
  };
};

export const getStaticProps: GetStaticProps = async ({ params }) => {
  const { category, page } = params;

  const limit = 2;
  let skip;
  if (JSON.parse(page) === 1) {
    skip = 0;
  } else {
    skip = JSON.parse(page) * limit - limit;
  }
  const blogs = await client.getEntries({
    content_type: "blog",
    limit: limit,
    skip: skip,
    "fields.category": category,
  });

  return {
    props: {
      page: JSON.parse(page),
      blogs: blogs.items,
    },
    revalidate: 1,
  };
};
```

I need to get access to dynamic category string from url in getStaticPaths so I can get the exact number of blogposts for that specific category but couldn't have access to anything at all .

How can I fix this problem ? Is there a workaround to fix this ?

Thanks in advance .

**Comment by icyJoseph** [comment]
Is it possible for you to get the categories list from your database? and then use those to in turn generate the blog posts? 

It should be possible to query your DB to get all categories, I think, and also leave the door open with `fallback:true` so Next.js can generate more pages, when more categories are created while the app is deployed.

If it's not possible to get all categories, perhaps, you'd need to collect all blog posts, take their category field, and unite all of that into a huge list. I'd say it is best to not render them all at build time, but, render the most visited at build time, and use `fallback:true` to render the rest as requested.

The reason why you can't access this path property in `getStaticPaths` is because, `getStaticPaths` is in charge of generating them in the first place.

**Reply by 1mehdifaraji** [reply]
I solved the issue by generating all categories pages in getStaticPaths and returned the final path like this :
```
let paths = [];

  newCategories.forEach((eachCategory: any) => {
    let pages;
    const howmany = eachCategory.posts.length / limit;
    pages = Math.ceil(howmany / 1) * 1;
    for (var i = 0; i < pages; i++) {
      paths.push({
        params: { page: `${i + 1}`, category: eachCategory.category },
      });
    }
  });

  return {
    paths,
    fallback: false,
  };
```
This way all blog categories pages are generated and returned in paths of getStaticPaths . 

Here is [stackoverflow](https://stackoverflow.com/questions/72667021/how-to-use-getstaticpaths-dynamically-in-next-js) answer as well

**Reply by icyJoseph** [reply]
Awesome, yeah that was pretty much the only option, generate them from a source inside `getStaticPath`, nobody asked but I refactored that a bit:

```tsx
import type { GetStaticPathsResult } from "next";

type PagePathResult = GetStaticPathsResult<{
  page: string;
  category: string;
}>["paths"];

export const getStaticPaths = () => {
  const limit = 100; // of course I made this one up
  const newCategories: Array<{ posts: []; category: string }> = []; // and this probably comes from a network call

  const paths = newCategories.reduce<PagePathResult>(
    (prev, currentCategory) => {
      const length = Math.ceil(currentCategory.posts.length / limit);

      const categoryPages = Array.from({ length }, (_, i) => ({
        params: {
          page: `${i + 1}`,
          category: currentCategory.category,
        },
      }));

      prev.push(...categoryPages);

      return prev;
    },
    []
  );

  return {
    paths,
    fallback: false,
  };
};

```
